File: 프로그래머스/3/60062. 외벽 점검/외벽 점검.cpp
#include <bits/stdc++.h>
using namespace std;

int dp[8][1 << 16];
const int INF = 987654321;
vector<int> w, d;
int N,ret;

int go(int idx, int visited) {
    if (visited == (1 << w.size()) - 1) return 0;
    if (idx == d.size()) return INF;

    int& ret = dp[idx][visited];
    if (ret != -1) return ret;
    ret = INF;

    for(int i=0;i<w.size();i++){
        if ((visited & 1 << i)) continue;
        int temp = visited;
        for(int j=0;j<w.size();j++) {
            int diff = (w[j] - w[i] + N) % N;
            if (diff <= d[idx])
                temp |= 1 << j;
        }
        ret = min(ret, go(idx + 1, temp) + 1);
    }
    

    return ret;
}


int solution(int n, vector<int> weak, vector<int> dist) {
    w = weak;
    d = dist;
    N = n;

    sort(d.rbegin(), d.rend());
    memset(dp, -1, sizeof(dp));

    ret = go(0, 0);

    return (ret==INF ? -1 : ret) ;
}

File: 프로그래머스/3/60062. 외벽 점검/외벽 점검.js
function changeStart(weak, start, n) {
  if (start === 0) return weak;

  let v = [];
  for (let i = start; i < weak.length; i++) {
    v.push(weak[i]);
  }
  for (let i = 0; i < start; i++) {
    v.push(weak[i] + n);
  }
  return v;
}

function solution(n, weak, dist) {
  const INF = 987654321;
  let ret = INF;

  dist.sort();

  do {
    for (let i = 0; i < weak.length; i++) {
      let newWeak = changeStart(weak, i, n);
      let idx = 0,
        flag = 0;
      let cur = newWeak[0] + dist[0];

      for (let j = 0; j < weak.length; j++) {
        if (cur < newWeak[j]) {
          if (idx + 1 === dist.length) {
            flag = 1;
            break;
          }
          cur = newWeak[j] + dist[++idx];
        }
      }
      if (!flag) ret = Math.min(ret, idx + 1);
    }
  } while (next_permutation(dist));

  return ret === INF ? -1 : ret;
}

// Helper function for generating permutations
function next_permutation(arr) {
  let i = arr.length - 1;
  while (i > 0 && arr[i - 1] >= arr[i]) i--;
  if (i <= 0) return false;

  let j = arr.length - 1;
  while (arr[j] <= arr[i - 1]) j--;
  let temp = arr[i - 1];
  arr[i - 1] = arr[j];
  arr[j] = temp;

  j = arr.length - 1;
  while (i < j) {
    temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
    i++;
    j--;
  }

  return true;
}
